### DISCUSSION TITLE
Automatically download and install updates in the future button disabled

### DISCUSSION THREAD

**Discussion by jzau** [sparkle-project/Sparkle#2199] [Q&A] [answered]
## Summary

![image](https://user-images.githubusercontent.com/2881736/178538429-e591469e-cc33-49c5-85a5-845444159ee3.png)
I setup a fresh new project, the only thing i put in the info.plist is the `SUPublicEDKey`. The button is disabled and not able to tick it.

```
updater = SPUUpdater(hostBundle: Bundle.main, applicationBundle: Bundle.main, userDriver: driver, delegate: self)
updater?.setFeedURL(URL(string: "http://localhost:8000/appcast.xml")!)
try? updater?.start()
updater?.checkForUpdates()
```

that's all the code i have in my test project. sandbox disabled.

## Possible Fix

I assume allowAutomaticUpdates is true by default? corrent me if i'm wrong, because i didn't put any other settings in the info.plist. 

```  
if (!self.allowsAutomaticUpdates) {
        if (showReleaseNotes) {
            // Fix constraints so that buttons aren't far away from web view when we hide the automatic updates check box
            NSLayoutConstraint *skipButtonToReleaseNotesContainerConstraint = [NSLayoutConstraint constraintWithItem:self.skipButton attribute:NSLayoutAttributeTop relatedBy:NSLayoutRelationEqual toItem:self.releaseNotesContainerView attribute:NSLayoutAttributeBottom multiplier:1.0 constant:12.0];
            
            [self.window.contentView addConstraint:skipButtonToReleaseNotesContainerConstraint];
            
            [self.automaticallyInstallUpdatesButton removeFromSuperview];
        } else {
            // Disable automatic install updates option if the developer wishes for it in Info.plist
            // If we are showing release notes, this button will be hidden instead
            self.automaticallyInstallUpdatesButton.enabled = NO;
        }
    }
```

## Version

in Podfile.lock
 - Sparkle (2.2.0)

Since i didn't set up anything about automatically update, I think it should give user an option to enable auto update instead of disabling this button. It would be great if someone can confirm if this is a bug. I can try to raise a PR if this is really a bug. 

Many thanks!

**Comment by zorgiepoo** [comment, marked as answer]
Take a look at [the programmatic setup](https://sparkle-project.org/documentation/programmatic-setup/) if you're just using Sparkle in a standard way and can use `SPUStandardUpdaterController`.

Please don't use `setFeedURL()` for new projects; the API documentation warns against this. Provide your feed using the [SUFeedURL key](https://sparkle-project.org/documentation/customization) in your Info.plist. Our [website documentation](https://sparkle-project.org/documentation/publishing/#channels) covers using channels and switching feeds if you need alternative appcasts.

`-checkForUpdates` is only intended for update checks that the users initiate themselves (e.g. from a menu bar item). The API documentation covers this.

Sparkle can automatically check for updates in the background otherwise based on a scheduled interval. By default, it will ask your user permission to check for new updates on the 2nd launch of your app. The button in your screenshot is disabled because Sparkle wasn't set to automatically check for updates yet which is a prerequisite for automatically downloading/installing updates.

Check our documentation page on [customizing Sparkle](https://sparkle-project.org/documentation/customization/). Generally you customize initial behavior in the Info.plist, not by code.

**Reply by jzau** [reply]
Thanks for your quick reply.

I got it, I got confused about automatically checking for updates and automatically installing updates. 

Since Sparkle wasn't set to automatically check for updates yet should we hide this button instead? As a user if i click the check for updates button for the first time I launch the app, I will get confused why can't I enable automatical install updates.

**Reply by zorgiepoo** [reply]
The button will be hidden if your update has release notes. Looking at the code in SUUpdateAlert.m I feel it was not hidden in this no-release-notes case for aesthetic / UI layout reasons. I suggest providing release notes anyway. Ignoring aesthetic and UI constraints, yes it technically should be hidden when no release notes are provided and Sparkle isn't set to automatically check for updates yet.

**Reply by jzau** [reply]
Yeah, it says that However when we don't show release notes, it looks ugly if the install buttons are too close to the description field.

hmm, maybe we should add more constraints in order to make the install buttons not too close to the description field rather than showing a disabled checkbox?

Anyway thanks for your explanation.

**Reply by zorgiepoo** [reply]
Indeed sounds like a good idea. Feel free to make a PR if you want.

https://github.com/sparkle-project/Sparkle/issues/2201

Also I noticed https://github.com/sparkle-project/Sparkle/issues/2200

**Reply by jzau** [reply]
sure, will do
